src/CBF/CBF.jl

```diff
@@ -0,0 +1,560 @@
+                i = div(1 + isqrt(8 * t.output_index - 7), 2)
+                j = t.output_index - div(i * (i - 1), 2)
+                push!(hcoord, (length(psd_side_dims), t.scalar_term.variable_index.value, i, j, t.scalar_term.coefficient))
+            end
+            k = 0
+            for i in 1:side_dim, j in 1:i
+                k += 1
+                row_constant = con_func.constants[k]
+                if !iszero(row_constant)
+                    push!(dcoord, (length(psd_side_dims), i, j, row_constant))
+                end
+            end
+            @assert k == MOI.output_dimension(con_func)
+        end
+
+        if !isempty(psd_side_dims)
+            println(io, "PSDCON")
+            println(io, length(psd_side_dims))
+            for side_dim in psd_side_dims
+                println(io, side_dim)
+            end
+            println(io)
+        end
+
+        if !isempty(hcoord)
+            println(io, "HCOORD")
+            println(io, length(hcoord))
+            for (psd_con_idx, var, i, j, coef) in hcoord
+                println(io, psd_con_idx - 1, " ", var - 1, " ", i - 1, " ", j - 1, " ", coef) # CBF indices start at 0.
+            end
+            println(io)
+        end
+
+        if !isempty(dcoord)
+            println(io, "DCOORD")
+            println(io, length(dcoord))
+            for (psd_con_idx, i, j, constant) in dcoord
+                println(io, psd_con_idx - 1, " ", i - 1, " ", j - 1, " ", constant) # CBF indices start at 0.
+            end
+            println(io)
+        end
+    end
+
+    return
+end
+
+# ==============================================================================
+#
+#   MOI.read_from_file
+#
+# The CBF file format (version 3) is described at
+# http://cblib.zib.de/doc/format3.pdf.
+#
+# ==============================================================================
+
+# Convert a pair of row and column indices of a symmetric matrix into a vector index for the row-wise lower triangle
+matrix_to_vector_idx(i::Int, j::Int) = (i < j) ? div((j - 1) * j, 2) + i : div((i - 1) * i, 2) + j
+
+function MOI.read_from_file(model::Model, filename::String)
+    if !MOI.is_empty(model)
+        error("Cannot read in file because model is not empty.")
+    end
+
+    scalar_vars = MOI.VariableIndex[]
+    psd_vars = Vector{MOI.VariableIndex}[]
+    obj_terms = MOI.ScalarAffineTerm{Float64}[]
+    obj_constant = 0.0
+    con_cones = Tuple{String, Int}[]
+    row_terms = Vector{MOI.ScalarAffineTerm{Float64}}[]
+    row_constants = Float64[]
+    psd_side_dims = Int[]
+    psd_cone_dims = Int[]
+    psd_row_terms = Vector{MOI.ScalarAffineTerm{Float64}}[]
+    psd_row_constants = Float64[]
+
+    open(filename, "r") do io
+        while !eof(io)
+            line = strip(readline(io))
+
+            # Skip blank lines and comments (CBF comments start with #).
+            if isempty(line) || startswith(line, "#")
```

Comment:
Is it possible to have in-line comments?

Reply:
not according to my interpretation of section 2.5 of [CBF version 3 doc ](http://cblib.zib.de/doc/format3.pdf)